core/src/main/java/org/axonframework/eventsourcing/EventSourcingRepository.java

```diff
@@ -204,7 +204,8 @@ public EventSourcingRepository(AggregateFactory<T> aggregateFactory, EventStore
      */
     @Override
     protected EventSourcedAggregate<T> doLoadWithLock(String aggregateIdentifier, Long expectedVersion) {
-        DomainEventStream eventStream = eventStore.readEvents(aggregateIdentifier);
+        DomainEventStream eventStream = eventStore.readEvents(aggregateIdentifier).filter(
+                e -> e.getType().equals(aggregateModel().type()));
```

Comment:
When introducing the new method readEvents (String aggregateId, Long expectedVersion) a possible override for the filtered version could be to call eventStore.readEvents(aggregateIdentifier, aggregateType). So if the EventStore could be extended with a readEvents method which also uses the type as suggested by @m1l4n54v1c there will be no performance penalty either. This would be the Axon2 compatible version. Which would need less adjustments instead of introducing a new class etc.

Reply:
Your obviously open to provide your idea of what would work best in this space. It is your PR after all. Looking forward to your commit @abbierwolf!

Reply:
For now I added the extra protected method and kept the FilteringDomainEventStream which should be enough to implement an filtered subclass myself. Implementing the readEvents with type filter in the EventStore can be done in multiple ways. Where do you discuss changes of interfaces? I am willing to spent some time in this but I need to know more about how you design your interfaces.

Reply:
Adding the `EventSourcingRepository#readEvents(String, Long)` function and keeping the `FilteringDomainEventStream` is exactly what I had in mind.

I didn't envision any adjustments to interface, let alone the `EventStore`, as this is a 'EventSourcing(Repository)' specific thing (for now at least). So like my first sentence points out, I think this is sufficient.

Maybe @abuijze has a suggestion for further adjustments on how to provide your use case and keep current use cases working.